libvast/native-plugins/compute.cpp

```diff
@@ -0,0 +1,152 @@
+        for (const auto& index :
+             caf::get<record_type>(layout).resolve_key_suffix(field,
+                                                              layout.name())) {
+          auto transformation
+            = [&](struct record_type::field field,
+                  std::shared_ptr<arrow::Array> array) noexcept
+            -> std::vector<std::pair<struct record_type::field,
+                                     std::shared_ptr<arrow::Array>>> {
+            auto execute_result
+              = function_->Execute({array}, function_options_.get(), nullptr);
+            // FIXME: be more lenient with error handling than asserting here.
+            VAST_ASSERT(execute_result.ok(),
+                        execute_result.status().ToString().c_str());
+            array = execute_result.MoveValueUnsafe().make_array();
+            field.type = type::from_arrow(*array->type());
+            return {{std::move(field), std::move(array)}};
+          };
+          indexed_transformations->second.push_back(
+            {index, std::move(transformation)});
+        }
+      }
+      std::sort(indexed_transformations->second.begin(),
+                indexed_transformations->second.end());
+    }
+    // Apply the transformations.
+    std::tie(layout, batch)
+      = transform_columns(layout, batch, indexed_transformations->second);
+    transformed_batches_.emplace_back(std::move(layout), std::move(batch));
+    return caf::none;
+  }
+
+  [[nodiscard]] caf::expected<std::vector<transform_batch>> finish() override {
+    return std::exchange(transformed_batches_, {});
+  }
+
+private:
+  /// Cache for transformed batches.
+  std::vector<transform_batch> transformed_batches_ = {};
+
+  /// Cache for the per-layout transformations.
+  std::unordered_map<type, std::vector<indexed_transformation>>
+    indexed_transformations_cache_ = {};
+
+  /// The paramters of the transformation.
+  std::shared_ptr<arrow::compute::Function> function_ = {};
+  std::unique_ptr<arrow::compute::FunctionOptions> function_options_ = {};
+  std::vector<std::string> fields_ = {};
+  std::vector<arrow::Datum> inputs_ = {};
+};
+
+// -- plugin ------------------------------------------------------------------
+
+class plugin final : public virtual transform_plugin {
+public:
+  caf::error initialize(data options) override {
+    // We don't use any plugin-specific configuration under
+    // vast.plugins.compute, so nothing is needed here.
+    if (caf::holds_alternative<caf::none_t>(options))
+      return caf::none;
+    if (const auto* rec = caf::get_if<record>(&options))
+      if (rec->empty())
+        return caf::none;
+    return caf::make_error(ec::invalid_configuration, "expected empty "
+                                                      "configuration under "
+                                                      "vast.plugins.compute");
+  }
+
+  [[nodiscard]] const char* name() const override {
+    return "compute";
```

Comment:
I find `compute` a bit too generic, so I'm open to ideas. Arrow calls the supported compute functions "element-wise" or "scalar" functions.

Reply:
I like "compute" for now. I also like "mutate" from dplyr land, which puts emphasis on the fact that you're messing with the data.

But the best way to approach this is figuring out how it will gel with the pipeline syntax, then work backwards to YAML configuration.